### PR TITLE
optimize buildings

### PR DESCRIPTION
**Pull request by BalaM314**
@buthed010203 please review this code
i still have that weird feeling that I did a big dumb

### CODE REVIEW

**Comment by BalaM314** [comment]
like i have compiled and tested and it seems to work
Profiles generated using VisualVM on Transcendence with max zoom in
Before:
image
After:
image
profiles may be inaccurate and/or stupid

File: core/src/mindustry/entities/comp/BuildingComp.java [hunk 1]
```diff
@@ -1001,23 +1002,36 @@ public boolean dump(Item todump){
 
         if(proximity.size == 0) return false;
 
+        if(todump == null){
+            //Prepare the list of items
+            dumpItems.clear();
+            for(int i = 0; i < content.items().size; i ++){
+                if(this.items.has(i)){
+                    dumpItems.add(content.item(i));
```
Comment (buthed010203) [inline core/src/mindustry/entities/comp/BuildingComp.java:1010]: Did we not just establish that `content.item(int)` is slow?

File: core/src/mindustry/world/modules/ItemModule.java [hunk 2]
```diff
@@ -21,7 +21,7 @@ public class ItemModule extends BlockModule{
     private static final Interval flowTimer = new Interval(2);
     private static final float pollScl = 20f;
 
-    protected int[] items = new int[content.items().size];
+    public int[] items = new int[content.items().size];//publicised in foos client
```
Comment (buthed010203) [inline core/src/mindustry/world/modules/ItemModule.java]: Is this actually used?

File: core/src/mindustry/entities/comp/BuildingComp.java (hunk 1 above)
Comment (BalaM314) [reply core/src/mindustry/entities/comp/BuildingComp.java:1010]: yes but this is outside both loops, which is called like 100x less

File: core/src/mindustry/world/modules/ItemModule.java (hunk 2 above)
Comment (BalaM314) [reply core/src/mindustry/world/modules/ItemModule.java]: no

File: core/src/mindustry/entities/comp/BuildingComp.java [hunk 3]
```diff
@@ -1001,23 +1002,36 @@ public boolean dump(Item todump){
 
         if(proximity.size == 0) return false;
 
+        if(todump == null){
+            //Prepare the list of items
+            dumpItems.clear();
```
Comment (buthed010203) [inline core/src/mindustry/entities/comp/BuildingComp.java]: Don't even bother clearing the seq, `dumpItems.size = 0` should do the trick

File: core/src/mindustry/entities/comp/BuildingComp.java (hunk 1 above)
Comment (buthed010203) [reply core/src/mindustry/entities/comp/BuildingComp.java:1010]: Still called a bunch but fair enough

File: core/src/mindustry/world/modules/ItemModule.java (hunk 2 above)
Comment (buthed010203) [reply core/src/mindustry/world/modules/ItemModule.java]: Probably want to revert it to keep compatibility then

**Pull request merged**